### ISSUE TITLE
Verifying action masking with ReinforceAgent/ActorPolicy

### ISSUE THREAD

**Issue by hornstenfilip** [tensorflow/agents#641]
Hi there,

Thanks for an awesome RL framework!

I'm trying to implement an RL agent with TF Agents for a custom OpenAI Gym environment and am having some trouble with action masking and the ReinforceAgent.

I'm using the following packages:

```
tf-agents-nightly=0.8.0.dev20210708
tf-estimator-nightly=2.6.0.dev2021062501
tf-nightly=2.7.0.dev20210706
tfp-nightly=0.14.0.dev20210630
```

The custom Gym environment is loaded into TF Agents and passed to a TensorFlow wrapper. Observations are dictionaries of the current state and corresponding mask.

The observation and action specs of the Python environment are

```
OrderedDict([('coordinates', BoundedArraySpec(shape=(10, 2), dtype=dtype('float32'), name='observation/coordinates', minimum=0.0, maximum=1.0)), ('mask', BoundedArraySpec(shape=(10,), dtype=dtype('int32'), name='observation/mask', minimum=0, maximum=1))])

BoundedArraySpec(shape=(), dtype=dtype('int64'), name='action', minimum=0, maximum=9)
```
and the TensorFlow environment has corresponding `BoundedTensorSpec` equivalents.

`validate_py_environment(py_env, episodes=5)` passes without errors.

An example TF environment observation looks like this:
```
tf.Tensor(
[[[0.82225746 0.21253699]
  [0.8870637  0.50194335]
  [0.9868757  0.5598946 ]
  [0.98240113 0.05240792]
  [0.39743358 0.29287595]
  [0.9909457  0.5113992 ]
  [0.28003615 0.12093569]
  [0.11129443 0.60465044]
  [0.59976405 0.84500533]
  [0.83713263 0.49189767]]], shape=(1, 10, 2), dtype=float32)
tf.Tensor([[1 0 1 1 1 1 1 1 1 1]], shape=(1, 10), dtype=int32)
```

I'm masking the actions by wrapping an [ActorDistributionNetwork](https://www.tensorflow.org/agents/api_docs/python/tf_agents/networks/actor_distribution_network/ActorDistributionNetwork) in a [MaskSplitterNetwork](https://www.tensorflow.org/agents/api_docs/python/tf_agents/networks/mask_splitter_network/MaskSplitterNetwork). I'm using the ReinforceAgent with standard policies and replay buffer.

```python
train_py_env = suite_gym.load(env_name)
train_env = tf_py_environment.TFPyEnvironment(train_py_env)


def splitter_fn(observation_and_mask):
    return observation_and_mask['coordinates'], observation_and_mask['mask']

actor_wrapped_net = actor_distribution_network.ActorDistributionNetwork(observation_spec['coordinates'], action_spec,
    fc_layer_params=fc_layer_params)

actor_net = mask_splitter_network.MaskSplitterNetwork(splitter_fn, actor_wrapped_net, passthrough_mask=True)


optimizer = tf.keras.optimizers.Adam(learning_rate=learning_rate)

train_step_counter = tf.Variable(0)

tf_agent = reinforce_agent.ReinforceAgent(
    train_env.time_step_spec(),
    train_env.action_spec(),
    actor_network=actor_net,
    optimizer=optimizer,
    normalize_returns=True,
    train_step_counter=train_step_counter)
tf_agent.initialize()

eval_policy = tf_agent.policy
collect_policy = tf_agent.collect_policy

replay_buffer = tf_uniform_replay_buffer.TFUniformReplayBuffer(
    data_spec=tf_agent.collect_data_spec,
    batch_size=train_env.batch_size,
    max_length=replay_buffer_capacity)
```

`actor_net.output_spec` is
```
DistributionSpec(builder=<class 'tensorflow_probability.python.distributions.categorical.Categorical'>, input_params_spec={'logits': TensorSpec(shape=(10,), dtype=tf.float32, name='CategoricalProjectionNetwork_logits')}, sample_spec=BoundedTensorSpec(shape=(), dtype=tf.int64, name='action', minimum=array(0, dtype=int64), maximum=array(9, dtype=int64)))
```

The agent does not seem to take illegal actions, but does not learn anything useful but instead only repeats the same actions except for the first and last actions because of restrictions imposed by the environment. Do I also need to create a custom policy? By default, `ReinforceAgent` uses an `ActorPolicy` as the `collect_policy` and a `GreedyPolicy` of the same policy as the `policy`. As far as I understand, the masked network is passed to the policies by the agent. I've tried to look at the distribution produced by the policies but `ReinforceAgent` doesn't have `emit_log_probability` like e.g. `DqnAgent`. The result of

```python
distribution(train_env.current_time_step())
```

for the `policy` is
```
PolicyStep(action=<tfp.distributions.DeterministicWithLogProb 'Deterministic' batch_shape=[1] event_shape=[] dtype=int64>, state=(), info=())
```

and for the `collect_policy` is
```
PolicyStep(action=<tfp.distributions.Categorical 'MaskSplitterNetwork_ActorDistributionNetwork_CategoricalProjectionNetwork_Categorical' batch_shape=[1] event_shape=[] dtype=int64>, state=(), info=())
```

Do I need to create a custom `ActorPolicy` and manually set `observation_and_action_constraint_splitter` or tinker with the `MaskedCategorical` distribution? The only source I could find is an [almost two-year-old issue](https://github.com/tensorflow/agents/issues/255#issuecomment-560497090) that I believe predates the creation of the `MaskSplitterNetwork`.

Thank you so much for any help!

**Comment by hornstenfilip** [comment]
I was able to fix it. One can check by accessing the action distribution of the policy by
```python
collect_policy.distribution(train_env.current_time_step()).action.log_prob((0,1,2,3,4)).numpy()
```

, stepping through the environment and making sure that the logits are indeed masked out.
